prql-compiler/src/parser/lexer.rs

```diff
@@ -0,0 +1,322 @@
+#![allow(dead_code)]
+
+use chumsky::prelude::*;
+use itertools::Itertools;
+
+use crate::ast::pl::*;
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Token {
+    Whitespace,
+    NewLine,
+
+    Ident(String),
+    Literal(Literal),
+
+    Interpolation(char, Vec<InterpolItem>),
+
+    // this contains 3 bytes at most, we should replace it with SmallStr
+    Control(String),
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum InterpolItem {
+    String(String),
+    Expr(String),
+}
+
+pub fn lexer() -> impl Parser<char, Vec<(Token, std::ops::Range<usize>)>, Error = Simple<char>> {
+    let whitespace = just('\t')
+        .or(just(' '))
+        .repeated()
+        .at_least(1)
+        .to(Token::Whitespace);
+
+    let new_line = just('\r').or_not().then(just('\n')).to(Token::NewLine);
+
+    let control_multi = just("->")
+        .or(just("=>"))
+        .or(just("=="))
+        .or(just("!="))
+        .or(just(">="))
+        .or(just("<="))
+        .or(just("and")) // TODO: negative lookahead for whitespace
+        .or(just("or")) // TODO: negative lookahead for whitespace
+        .or(just("??"))
+        .map(|x| x.to_string())
+        .map(Token::Control);
```

Comment:
I'd recommend [`choice`](https://docs.rs/chumsky/latest/chumsky/primitive/fn.choice.html) for `or` chains like this. It's faster to compile and a bit easier to maintain.